core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -66,7 +66,33 @@ sealed trait ConfigCursor {
    *         otherwise.
    */
   def asListCursor: Either[ConfigReaderFailures, ConfigListCursor] =
-    castOrFail(LIST, Set(LIST), _.asInstanceOf[ConfigList]).right.map(ConfigListCursor(_, pathElems))
+    if (isUndefined) {
+      failed(KeyNotFound.forKeys(path, Set()))
+    } else {
+      val ve = value.valueType() match {
+        case ConfigValueType.OBJECT =>
+          val ll =
+            value.asInstanceOf[ConfigObject].asScala.foldLeft[Either[List[String], List[(Int, ConfigValue)]]](Right(Nil)) {
+              case (acc, (str, v)) =>
+                (acc, Try(str.toInt)) match {
+                  case (Right(a), Success(b)) => Right(a :+ (b -> v))
+                  case (Left(keys), Failure(_)) => Left(keys :+ str)
```

Comment:
`:+` has linear complexity. Since we don't need to preserve the order here (`ConfigObject` is unordered), it's better to use `str :: keys`. Same above.

Reply:
Agreed. Fixed that in 6fb1c520eff4b7333f3df2fef50a47764e6c4c7c.